### ISSUE TITLE
Mcrouter performance problem

### ISSUE THREAD

**Issue by AlexKapustin** [facebook/mcrouter#96]
Hi!
Too slow performance of mcrouter.

TestScript:
```
<?php

$keys = 'ITEM 9ff4b66af0f05309d2531dd2c75bf23bbdbfd18d [219 b; 1447397847 s]
ITEM a8812e8757234fbc6261b1050bfd4474a2c0c846 [216 b; 1447397847 s]
ITEM 9d6a00020b23479914a6d810171291de58217b31 [204 b; 1447397847 s]
ITEM 1e6bae6b0763b3424c3c26da5a6b2feb562e2748 [205 b; 1447397847 s]
ITEM 00fe64bd56546f3451b6f45398e2c284e27f171d [221 b; 1447397847 s]
ITEM cdae1d16f462cbcee90603cddeb91c1a2bc76bd7 [215 b; 1447397847 s]
ITEM 6a9b32414fcbd7a7e2d8da1ec6a50f3bc3c98f9c [239 b; 1447397847 s]
ITEM 4b711b28e7ce230d8af694886a2c706d74a6a102 [241 b; 1447397847 s]
ITEM 27c10605925616592291bbf987f71cb83c571c9a [206 b; 1447397847 s]
ITEM d6becf1007e45e79afef86f84219203d0f945015 [209 b; 1447397847 s]';

$keys = explode("\n", $keys);
foreach ($keys as &$key) {
    $key = explode(' ', $key);
    $key = $key[1];
}

$mc = new Memcached();
$mc->addServer('localhost', 2314);

$start = microtime(1);
for ($i = 0; $i < 10000; $i++) {
    foreach ($keys as $key) {
        $data = $mc->get($key);
    }
}

echo 'Mcrouter: ', (microtime(1) - $start), PHP_EOL;

$mc = new Memcached();
$mc->addServer('cache1.lzd', 11214);

$start = microtime(1);
for ($i = 0; $i < 10000; $i++) {
    foreach ($keys as $key) {
        $data = $mc->get($key);
    }
}
echo 'Memcached: ', (microtime(1) - $start), PHP_EOL;
```

My output:
```
Mcrouter: 31.141890048981
Memcached: 0.50829887390137
```

If we replace `foreach` by `getMulti` it's better, but still too slow:
```
Mcrouter: 5.2232239246368
Memcached: 0.28168606758118
```

Mcrouter config:

```
{
  "pools": {
    "pool-foo": {
      "servers": [
        "cache1:11312",
        "cache2:11312",
        "cache3:11312",
        "cache4:11312",
        "cache5:11312",
        "cache6:11312",
        "cache7:11312",
        "cache8:11312",
        "cache9:11312",
        "cache10:11312",
        "cache11:11312",
        "cache12:11312",
        "cache13:11312",
        "cache14:11312",
        "cache15:11312",
        "cache16:11312"
      ]
    }
  },
  "route": {
    "type": "OperationSelectorRoute",
    "operation_policies": {
      "delete": "AllFastestRoute|Pool|pool-foo",
      "set": "AllFastestRoute|Pool|pool-foo",
      "add": "AllFastestRoute|Pool|pool-foo",
      "incr": "AllFastestRoute|Pool|pool-foo",
      "decr": "AllFastestRoute|Pool|pool-foo",
      "get": {
        "type": "FailoverRoute",
        "children": [
          "RandomRoute|Pool|pool-foo",
          "LatestRoute|Pool|pool-foo"
        ]
      },
      "gets": {
        "type": "FailoverRoute",
        "children": [
          "RandomRoute|Pool|pool-foo",
          "LatestRoute|Pool|pool-foo"
        ]
      }
    }
  }
}
```

**Comment by solidfoxrock** [comment]
Hi, just want to add some more information (I and @AlexKapustin are in the same team):
- we're using 10Gbps fiber network
- we're using a bit old version of mcrouter, commit 8303e73 (since July 6th)
- we're running mcrouter with 1 proxy thread
- our command options: `--use-asynclog-version2 --probe-timeout-initial 1000 --cluster-pools-timeout 500 --maximum-soft-tkos 15 --timeouts-until-tko 10`
- during high traffic, our mcrouter (running locally on web servers) handled around 10k-20k requests/sec from local php-fpm, and the response time from mcrouter became drastically slow

The simple tests above show that mcrouter adds a lot of latency with a simple replicated setup, which really surprised us. We're very curious about the performance of mcrouter in Facebook, maybe there are some tuning techniques that we can apply to improve the performance? Or maybe we're doing something wrong here.

Thanks,
Kien

**Comment by amcrn** [comment]
(note: i'm not from fb, but chiming in with a quick tip until the experts arrive)

@solidfoxrock @AlexKapustin in many cases you want num_proxies to be equal to the number of vcpus. see the comment here: https://github.com/facebook/mcrouter/issues/25#issuecomment-77008864

**Comment by alikhtarov** [comment]
The biggest performance impact would be due to the fact that internally we use jemalloc instead of standard glibc malloc - so we optimize everything with that in mind. We committed a few changes to our build scripts in the most recent version which will use jemalloc by default, please retry the test with that release.
As a workaround, bumping the number of threads as @amcrn suggested should help (in the newest release, we set it to the number of CPU cores by default).

**Comment by solidfoxrock** [comment]
Thank you @amcrn and @alikhtarov! We have solved our problem with bumping the number of threads. 

It's not that we were not aware of the fact that mcrouter is running with single thread as default, but at that time our benchmark tests showed same results between single thread and multiple threads. Turns out it was because we were running too many services on the same server so mcrouter could not utilize and show benefits in multithread mode. Now we keep the setup minimal and it shows very good result.

Thanks all.
